internal/security/secretstore/certs.go

```diff
@@ -0,0 +1,220 @@
+/*******************************************************************************
+ * Copyright 2019 Dell Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ * @author: Daniel Harms, Dell
+ * @author: Tingyu Zeng, Dell
+ *******************************************************************************/
+
+package secretstore
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	url2 "net/url"
+
+	"github.com/edgexfoundry/edgex-go/internal"
+)
+
+type CertCollect struct {
+	Pair CertPair `json:"data"`
+}
+
+type CertPair struct {
+	Cert string `json:"cert,omitempty"`
+	Key  string `json:"key,omitempty"`
+}
+
+type Certs struct {
+	client    internal.HttpCaller
+	certPath  string
+	tokenPath string
+}
+
+type auth struct {
+	Token string `json:"root_token"`
+}
+
+func NewCerts(caller internal.HttpCaller, certPath string, tokenPath string) Certs {
+	return Certs{client: caller, certPath: certPath, tokenPath: tokenPath}
+}
+
+func (cs *Certs) certPathUrl() (string, error) {
+	baseURL, err := url2.Parse(Configuration.SecretService.GetSecretSvcBaseURL())
+	if err != nil {
+		e := fmt.Errorf("error parsing secret-service url.  check server and port properties")
+		LoggingClient.Error(e.Error())
+		return "", err
+	}
+
+	certPath, err := url2.Parse(cs.certPath)
+	if err != nil {
+		e := fmt.Errorf("error parsing secret-service certpath.  check certpath property")
+		LoggingClient.Error(e.Error())
+		return "", err
+	}
+
+	fullUrl := baseURL.ResolveReference(certPath)
+	return fullUrl.String(), nil
+}
+
+func (cs *Certs) retrieve(t string) (*CertPair, error) {
+	url, err := cs.certPathUrl()
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		e := fmt.Errorf("error creating http request: %v", err.Error())
+		LoggingClient.Error(e.Error())
+		return nil, e
+	}
+
+	req.Header.Set(VaultToken, t)
+	resp, err := cs.client.Do(req)
+	if err != nil {
+		e := fmt.Errorf("failed to retrieve the proxy cert on path %s with error %s", cs.certPath, err.Error())
+		LoggingClient.Error(e.Error())
+		return nil, e
+	}
+	defer resp.Body.Close()
+
+	cc := CertCollect{}
+
+	if resp.StatusCode == http.StatusNotFound {
+		e := fmt.Errorf("proxy cert pair NOT found in secret store @/%s, status: %s", cs.certPath, resp.Status)
+		LoggingClient.Info(e.Error())
```

Comment:
should we return here from the function since we didn't find the token in the store and as such there won't be anything to decode later?